textbook/2_grouping.qmd

```diff
@@ -0,0 +1,994 @@
+
+Lastly, plotting data points along a single axis can lead to  overlap. That's where **jittering** comes to the rescue. Jittering introduces slight variation, or 'jitter', in the position of data points in each lane, making them easier to see. 
+
+::: {.column-margin}
+
+**Jitterbug**
+
+Back in the 1930s, the jitterbug was a lively, often wild, dance style that matched the fast-paced, unpredictable energy of swing music. In data visualization, jittering serves a similar purpose: to add a touch of randomness to points in a plot.
+
+![](/img/640px-Jitterbug_Wolcott_FSA.jpg)
+
+:::
+
+Let's examine how jittering works.
+
+::: {.panel-tabset}
+
+# Altair
+
+In Altair, we create a random calculated variable called `jitter` by appending the `.transform_calculate()` method to our chart. We use our newly created `jitter` variable, encoding it as the `yOffset` of each data point using `yOffset='jitter:Q'`. This shifts the `yOffset` of each point by a random value. 
+
+The `:Q` simply tells Altair that `jitter` is a continuous quantitative variable.
+
+```{python}
+alt.Chart(income_df).mark_point().encode(
+    x='percentage_heavy_drinkers',
+    y='region',
+    color='region',
+    yOffset="jitter:Q"
+).transform_calculate(
+    jitter="random()"
+)
+```
+
+# ggplot
+
+In ggplot2, you can achieve the same effect with jittering by using the `position_jitter()` function. Jitter is directly applied to the marks through the position argument within `geom_point()`.
+
+```{r fig.width=8, fig.height=6}
+ggplot(income_df, aes(
+    x=percentage_heavy_drinkers, 
+    y=region, 
+    color=region
+)) + 
+geom_point(position = position_jitter(width = 0, height = 0.2))
+```
+
+:::
+
+
+::: {.callout-caution}
+Jittering can be confusing! It can easily be mistaken for a scatterplot with true variation in the underlying data.
+:::
+
+## Faceting
+
+Faceting takes the concept of grouping to the next level by generating multiple small charts, each dedicated to a subset of the data. It is particularly effective for highlighting distinct trends and patterns in each subset, as well as identifying extreme values or outliers within groups. And unlike grouping by position, faceting keeps both the x and y axis free for other variables.
+
+Let's check out the syntax:
+
+::: {.panel-tabset}
+
+# Altair
+
+In Altair, faceting by World Bank income group is straightforward. We simply append `.facet('income_group')` to our chart. Additionally, we can control the layout by setting the number of columns using the `columns` argument:
+
+```{python}
+alt.Chart(income_df).mark_point().encode(
+    x='percentage_heavy_drinkers',
+    y='life_expectancy_at_50',
+).facet(
+    'income_group',
+    columns=2
+)
+```
+
+# ggplot
+
+Facetting works via `facet_wrap`. By default it tries to create an even num of cols and rows, rather than putting all plot in one row like Altair. However, you can specify the desired number of columns with the `ncol` argument.
```

Comment:
```suggestion
To facet by income group, we can append `.facet_wrap(vars(income_group))` to our chart. The `vars` function is required for `facet_wrap` to understand that we are referring to column names just like `aes` is required when typing out column names for the different visual channels.

When we are faceting the xy-axes are the same for all the grouped facets/subcharts. The default layout in ggplot is to achieve and even number of columns and rows. This is good when we are low on space, but otherwise it is preferred to spread the charts out horizontally in a single row if the primary purpose was to compare the charts' y-axis, since they would all be aligned, or vertically in a single column if we are the most interested in comparing the groups in terms of the values on the x-axis. Here, we don't have an explicit primary comparison, but we show how to change from the default horizontal layout by using the `ncol` parameter.
```

`vars` is recommend over ~ in general nowadays, neither is very intuitive in my opionion.